hastexo/tasks.py

```diff
@@ -135,6 +135,57 @@ def run(self, **kwargs):
         for key, value in kwargs.items():
             setattr(self, key, value)
 
+        # If a time limit is set for using labs,
+        # check how much time learner has already spent
+        settings = get_xblock_settings()
+        lab_usage_limit = settings.get("lab_usage_limit", None)
+
+        policy_warn_message = None
+        if lab_usage_limit:
+            # don't look further than 365 days for lab launch
+            cutoff = timezone.now() - timezone.timedelta(days=365)
+
+            # get all learner stacks across the platform
+            stacks = Stack.objects.filter(learner__id=self.learner_id
+                ).filter(launch_timestamp__gt=cutoff)
+
+            # add up total time spent on labs in seconds
+            total_time_spent = 0
+            for stack in stacks:
+                # A stacklog is saved each time the stack status changes,
+                # we only need to look at log entires with SUSPEND_COMPLETE
+                # status to get all records of active lab sessions with
+                # launch and suspend timestamps.
+                stacklog = StackLog.objects.filter(stack_id=stack.id
+                    ).filter(status="SUSPEND_COMPLETE")
+
+                # add up time spent on labs for one stack in seconds
+                time_spent = 0
+                for logentry in stacklog:
+                    time = (
+                        logentry.suspend_timestamp - logentry.launch_timestamp)
+                    time_spent += time.total_seconds()
+                total_time_spent += time_spent
+
+            if total_time_spent > lab_usage_limit:
+                policy = settings.get("lab_usage_limit_breach_policy",
+                                      "").lower()
+
+                if policy:
+                    logger.error(
+                        f'Learner {stacks[0].learner.email} has gone over '
+                        'the lab usage limit!')
+
+                if policy == 'block':
+                    raise Exception(
```

Comment:
I'm not sure how I feel about this being a generic Exception and being displayed as such to the learner; the fact that the pop-up reads `Exception("<message>")` isn't particularly pretty. I'm not sure what the best option here; should this perhaps be a subclass of `LaunchStackFailed`?

Reply:
Right, yes, I started out with `LaunchStackFailed` but ran into some errors how javascript handled it and switched to the generic one for testing. I'll take another look at that now :)

Reply:
One more thought on this: when blocked, the stack `status` ends up being `LAUNCH_ERROR`. I guess that's OK, we just need to make sure that such stacks get correctly reaped at the end of their `delete_age`. Since `LAUNCH_ERROR` is _not_ in the `dont_delete` list in [ReaperJob](https://github.com/hastexo/hastexo-xblock/blob/master/hastexo/jobs.py#L112) I would expect this to behave correctly and no changes are needed; do you agree?